### PR TITLE
Introduce basic linting infrastructure

### LINKED ISSUE #29: Define `.clang-format` or equivalent
`.clang-format` file defines the formatting rule for cpp files and allow modern tool chain to lint & auto-format 

This file has already been defined for [Optional26](https://github.com/beman-project/Optional26/blob/main/.clang-format) and [net29](https://github.com/beman-project/net29/blob/main/.clang-format) (note: those files are equivalent in terms of content).

### PR DESCRIPTION
**Pull request by wusatosi**
Introduce pre-commit based linting infrastructure:

- pre-commit **formats** extra white space, extra line at end-of-file, C++ files, and CMake files. It **flags** linting error for Markdown.
- CI does not push formatted change to pr/ main branch.
- Pre-commit is portable (even with `clang-format`) and does not pollute global scope, thus no virtual environment facing files (like `requirements.txt`) is included. This is tested with an empty GitHub Codespace.

Action Items:
- [x] setup a basic pre-commit infrastructure
- [x] setup a basic clang-format file, prefer discussion at #29 .
- [x] setup clang-format based formatting, see: 92b62736171fa9fd7692fc555f483ec85f0cdb5e for format diff
~~setup clang-tidy linting~~: may be out of scope, see discussion below: https://github.com/beman-project/exemplar/pull/34#discussion_r1784958937
- [x] setup cmake formatting (use [`gersemi`](https://github.com/BlankSpruce/gersemi), see: f3557fc485517922a656653a38c0a1c3fa662b60 for format diff)
- [x] setup markdown linting (via [`markdown-cli`](https://github.com/igorshubovych/markdownlint-cli), see: 1ca633837d4ecca08c58fb84c098cff4ed193fef for format diff)
- [x] setup ci check for linting
- [x] update project so it passes the linting test
- [ ] Update documentation (README)

Reference implementation: https://github.com/beman-project/optional26/pull/52
- Note that this PR diverges with reference implementation greatly:
- There's less "hooks" (linting rules)
- Uses different & updated tool chains
- The `.clang-format` is copied from [Optional 26](https://github.com/beman-project/Optional26/blob/main/.clang-format) & [net29](https://github.com/beman-project/net29/blob/main/.clang-format) with no modification.

Closes: #29 .

### CODE REVIEW

File: include/beman/exemplar/identity.hpp
Comment (JeffGarland) [inline include/beman/exemplar/identity.hpp:33]: I realize this is just a formatting PR, but I'm wondering if we actually need this macro for Beman? This would imply support for something prior to c++11 which really isn't our aim here.

Comment (wusatosi) [reply include/beman/exemplar/identity.hpp:33]: Yes this is a formatting pr! I am just presenting a preview of the formatting that is going to be adopted.

This is a good point! There's also a macro test for `__cpp_lib_transparent_operators` at testing implying C++14.

File: src/beman/exemplar/identity.t.cpp
Comment (JeffGarland) [inline src/beman/exemplar/identity.t.cpp:48]: That's a weird format rule - I suppose that's: AlignConsecutiveAssignments.  Anyway, I'd never write that -- maybe we can add a space after the container assignment so this is more normal spacing.

Comment (wusatosi) [reply src/beman/exemplar/identity.t.cpp:48]: Agreed!

File: include/beman/exemplar/identity.hpp
Comment (JeffGarland) [reply include/beman/exemplar/identity.hpp:33]: yes, well I guess I could see some library want c++14 support, but 98 -- not really.  I suggest we move the macro at 33 as a drive by fix on this PR.

Comment (wusatosi) [reply include/beman/exemplar/identity.hpp:33]: That's fair! After the weekly sync yesterday I think I will expand the scope of this pr out to be "Introduce general pre-commit + linting", haven't got around to updating the title and description, so I will need to work on this for a bit, I will open a separate pr specifically deleting this line.

Comment (JeffGarland) [reply include/beman/exemplar/identity.hpp:33]: Ok I'm good with that

File: src/beman/exemplar/identity.t.cpp
Comment (wusatosi) [reply src/beman/exemplar/identity.t.cpp:48]: Do you mind re-posting this observation to #29 ? I personally find this janky as well. I expect we will update `.clang-format` after we fight it out in a (or many) weekly sync afterwards, this pr aims to provide a implementation as close as possible to that in the optional26 project.

File: include/beman/exemplar/identity.hpp
Comment (JeffGarland) [reply include/beman/exemplar/identity.hpp:33]: As another note you'll see in optional26 PR the unadorned use of constexpr  https://github.com/beman-project/optional26/pull/64/files

Comment (wusatosi) [reply include/beman/exemplar/identity.hpp:33]: `unadorned`, cool new vocab for me!

I think I will actually hold creating a pr regarding this in hope of `clang-tidy` flagging it.

File: src/beman/exemplar/identity.t.cpp
Comment (JeffGarland) [reply src/beman/exemplar/identity.t.cpp:48]: done

File: include/beman/exemplar/identity.hpp
Comment (wusatosi) [reply include/beman/exemplar/identity.hpp:33]: May be bundled into #38

File: .cmake-format.json
Comment (wusatosi) [inline .cmake-format.json:15]: This looks rather weird to me, will need to investigate on this file.

File: .pre-commit-config.yaml
Comment (wusatosi) [inline .pre-commit-config.yaml:10]: These are basic pre-commit hooks, seem reasonable to me to add in

File: .pre-commit-config.yaml
Comment (wusatosi) [inline .pre-commit-config.yaml:21]: This is a manual trigger of the in-place formatting

File: .pre-commit-config.yaml
Comment (wusatosi) [inline .pre-commit-config.yaml:17]: Needs examples folder as well

File: .pre-commit-config.yaml
Comment (wusatosi) [inline .pre-commit-config.yaml:37]: Latest release in 4 years. We might want to find an alternative.

Comment (wusatosi) [reply .pre-commit-config.yaml:37]: This project is a configuration placement of [`cmake-format`](https://github.com/cheshirekow/cmake_format). 
Which seem to be abandoned, see: https://github.com/cheshirekow/cmake_format/issues/340

File: .pre-commit-config.yaml
Comment (wusatosi) [inline .pre-commit-config.yaml:23]: Project have not been updated been updated for 3 years. 

Seem not to be actively maintained, latest respond to issue tracker was in 2023.
This seem to just be a wrapper around calling clang-format, but an alternative maybe preferred.

File: .pre-commit-config.yaml
Comment (wusatosi) [reply .pre-commit-config.yaml]: omitted, let's find something better. Issue raised to optional project at: https://github.com/beman-project/optional26/issues/67

File: .cmake-format.json
Comment (wusatosi) [reply .cmake-format.json]: File removed.

File: .pre-commit-config.yaml
Comment (wusatosi) [inline .pre-commit-config.yaml:30]: This doesn't work, we will need to pass in project structure information to clang-tidy:

```
clang-tidy...............................................................Failed
- hook id: clang-tidy
- exit code: 1

/home/bradwu/Desktop/exemplar/examples/identity_direct_usage.cpp:3:10: error: 'beman/exemplar/identity.hpp' file not found [clang-diagnostic-error]
    3 | #include <beman/exemplar/identity.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "examples/identity_direct_usage.cpp"
No compilation database found in /home/bradwu/Desktop/exemplar/examples or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
1 error generated.
Error while processing /home/bradwu/Desktop/exemplar/examples/identity_direct_usage.cpp.
Found compiler error(s).
/home/bradwu/Desktop/exemplar/src/beman/exemplar/identity.t.cpp:3:10: error: 'beman/exemplar/identity.hpp' file not found [clang-diagnostic-error]
    3 | #include <beman/exemplar/identity.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "src/beman/exemplar/identity.t.cpp"
No compilation database found in /home/bradwu/Desktop/exemplar/src/beman/exemplar or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
1 error generated.
Error while processing /home/bradwu/Desktop/exemplar/src/beman/exemplar/identity.t.cpp.
Found compiler error(s).
```

We will need to provide compilation database and "include" directive (both for header and gtest). 
This means enable `CMAKE_EXPORT_COMPILE_COMMANDS` in cmake and provide `-Iinclude/` (as well gtest equivalent, which depends on the build directory).
A better approach maybe integrating clang-tidy into cmake build infrastructure as a build target and invoking that target in pre-commit.

Needs help here @bretbrownjr @camio @steve-downey @neatudarius

File: .pre-commit-config.yaml
Comment (wusatosi) [inline .pre-commit-config.yaml:21]: This is replaced with:
https://github.com/beman-project/exemplar/blob/7c2ee2821bfe0d36a0f60333b104dc4e3a7353bf/.pre-commit-config.yaml#L27

**Comment by wusatosi** [comment]
`markdownlint` doesn't provide a auto-format functionality, which would be annoying. 

Initial error report:
```
README.md:89: MD007 Unordered list indentation
README.md:13: MD012 Multiple consecutive blank lines
README.md:22: MD012 Multiple consecutive blank lines
README.md:194: MD012 Multiple consecutive blank lines
README.md:9: MD013 Line length
README.md:16: MD013 Line length
README.md:20: MD013 Line length
README.md:35: MD013 Line length
README.md:134: MD013 Line length
README.md:234: MD013 Line length
README.md:248: MD013 Line length
README.md:270: MD013 Line length
README.md:276: MD013 Line length
README.md:289: MD013 Line length
README.md:80: MD022 Headers should be surrounded by blank lines
README.md:116: MD031 Fenced code blocks should be surrounded by blank lines
README.md:95: MD033 Inline HTML
README.md:96: MD033 Inline HTML
README.md:112: MD033 Inline HTML
README.md:113: MD033 Inline HTML
README.md:122: MD033 Inline HTML
README.md:123: MD033 Inline HTML
README.md:143: MD033 Inline HTML
README.md:144: MD033 Inline HTML
README.md:203: MD033 Inline HTML
README.md:204: MD033 Inline HTML
README.md:231: MD033 Inline HTML
README.md:232: MD033 Inline HTML
README.md:244: MD033 Inline HTML
README.md:245: MD033 Inline HTML
README.md:265: MD033 Inline HTML
README.md:266: MD033 Inline HTML
README.md:284: MD033 Inline HTML
README.md:285: MD033 Inline HTML
README.md:234: MD034 Bare URL used
```

I looked over all the report, we need to at least omit `MD033 Inline HTML` and `MD034 Bare URL used`.

For `MD033`, we need `<details>` and `<summary>` to provide fold-able snippets.

For `MD034`, I think the linter is having trouble parsing markdown inside of `<details>`.

https://github.com/beman-project/exemplar/blob/0ac15db802577211f477a707ea6c0ce3acfc2075/README.md?plain=1#L234

**Comment by steve-downey** [comment]
For CI I have a strong preference for not having the linter make changes. Most of them make mistakes. `clang-format` is a rare exception, and even it can produce nonsense on language forms it doesn't understand yet.

**Comment by wusatosi** [comment]
Markdown error report (in addition to `--disable=MD033`, which flags use of `<details>`):
```
README.md:9:81 MD013/line-length Line length [Expected: 80; Actual: 296]
README.md:13 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
README.md:16:81 MD013/line-length Line length [Expected: 80; Actual: 208]
README.md:20:81 MD013/line-length Line length [Expected: 80; Actual: 114]
README.md:22 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
README.md:35:81 MD013/line-length Line length [Expected: 80; Actual: 100]
README.md:134:81 MD013/line-length Line length [Expected: 80; Actual: 206]
README.md:234:81 MD013/line-length Line length [Expected: 80; Actual: 202]
README.md:248:81 MD013/line-length Line length [Expected: 80; Actual: 127]
README.md:270:81 MD013/line-length Line length [Expected: 80; Actual: 126]
README.md:276:81 MD013/line-length Line length [Expected: 80; Actual: 165]
README.md:2[89](https://github.com/beman-project/exemplar/actions/runs/11149346248/job/30988022469?pr=34#step:4:94):81 MD013/line-length Line length [Expected: 80; Actual: 204]
```
This looks reasonable to me.

**Comment by wusatosi** [comment]
> For CI I have a strong preference for not having the linter make changes. Most of them make mistakes. `clang-format` is a rare exception, and even it can produce nonsense on language forms it doesn't understand yet.

Current CI does not push format changes to pr/ main branch, I would be interested in putting auto clang-format in but that's out of the scope of this CI.

**Comment by wusatosi** [comment]
Besides updating README, all configuration should be done. Opening for comments and reviews.

File: .pre-commit-config.yaml
Comment (wusatosi) [reply .pre-commit-config.yaml:23]: Moved clang-format to [`an actively maintained clang-format package`](https://github.com/pre-commit/mirrors-clang-format) in fc4d919a02a4f43b9e335ae623a80bb035c1f592

Comment (wusatosi) [reply .pre-commit-config.yaml:23]: No alternative for `clang-tidy`, but it have more problems to worry about, see below.

File: .pre-commit-config.yaml
Comment (wusatosi) [reply .pre-commit-config.yaml]: Very tempted to exclude this change from this pr.

File: .pre-commit-config.yaml
Comment (camio) [inline .pre-commit-config.yaml:31]:
```suggestion
```

I think we can remove this given the problems you've been seeing.

**Review by camio — APPROVED** [review]
This PR looks great to me.

I've used markdownlint-cli in the past and it worked well. It's the first time I've used gersemi to format CMake and while the output was surprising in a couple instances, the fact it is automated overshadows this.

I think this should be merged and our style czar @JeffGarland can provide an improved .clang-format later.

Our `README.md` should include instructions on pre-commit, but it should be clear that this is only for developers intending to make changes to this repository (those building for use elsewhere don't need it). The [Carbon pre-commit instructions](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md#running-pre-commit) are fairly decent.

@wusatosi, let me know if you want me to merge this now (and leave documentation to another PR) or if I should hold off.

File: .pre-commit-config.yaml
Comment (camio) [reply .pre-commit-config.yaml]: It's perfectly reasonable to not include clang-tidy as part of this PR. I'm also fine with it being part of a CMake build.

Comment (wusatosi) [reply .pre-commit-config.yaml]: Cool. I will look into clang-tidy later.

File: .pre-commit-config.yaml
Comment (wusatosi) [reply .pre-commit-config.yaml:31]: Thanks for the review.

The main reason for choosing `gersemi` is because it seems to be the only CMake formatter I found that is being actively maintained. We will need to keep an eye on these tools as they still haven't released a major version yet.

Thanks for the link to the carbon pre-commit instructions, I don't have experience writing good documentations so this is very helpful.
The reason I have not written the documentation is I am waiting for more review on the tool chain, I can write the documentation after more maintainer agrees on the contents of this pr. I would be open to putting the documentation in a separate pr.

File: .pre-commit-config.yaml
Comment (wusatosi) [reply .pre-commit-config.yaml]: resolve to delegate to future pr

**Comment by wusatosi** [comment]
Needs to include: 
https://github.com/pre-commit/action/issues/7#issuecomment-1251300704

And merge main.

**Review by steve-downey — APPROVED** [review]
:+1: 

And I'd like to apply some of this to optional26 as well, soon.

File: .pre-commit-config.yaml
Comment (linusboehm) [inline .pre-commit-config.yaml:5]: any reason to not use the most recent version here? (v5.0.0)

Comment (wusatosi) [reply .pre-commit-config.yaml:5]: Ops! Nice catch!

**Comment by wusatosi** [comment]
Weird, pr actions are not triggered anymore. GitHub must be down or smt

**Comment by wusatosi** [comment]
I don't think pre-commit's action's caching is working.
I have never seen it reuse the previous environment (always cache-miss).
I will keep an eye on this after the pr is merged and file an issue upstream later if needed.

**Comment by wusatosi** [comment]
@camio  This pr is ready to merge now, I don't want to block progress of #46 as we have overlapping config file.

I can file a separate pr to add documentation.

**Review by steve-downey — APPROVED** [review]
+1

File: include/beman/exemplar/identity.hpp
Comment (wusatosi) [reply include/beman/exemplar/identity.hpp:33]: Fix in #55

**Pull request merged**
